### ISSUE TITLE
unBalance alters folder dates on "collect" function

### ISSUE THREAD

**Issue by Attidudinal** [jbrodriguez/unbalance#9]
When 'collecting' folders (an invaluable function that saves LOTS of time over a 'manual' operation) a problem is created by the handling of date/time changes to the folder(s) in question...
Obviously, the folder that is 'going away' is not as issue as it no longer exists, however, the "target" folder (target drive specification) treats the folder AS IF IT WERE CREATED ANEW; causing all kinds of repercussions for date sensitive applications.
Changing the "accessed" date, or "modified" date (not really a good idea as it necessitates changing it back) is one thing; changing the "created" date is VERY PROBLEMATIC.
It was in fact created on the original date of its creation (it has not logically been recreated) as it existed before and still exists (only with additional contents) AND
By the time the "created" date has ALSO been altered (to today's date) along with all the other dates in question, IT RENDERS IT A PRACTICAL IMPOSSIBLITY TO RECTIFY ANY DATES (lacking any reference).
One would have to create a 'print out' of all affected files in advance of any collection which is a lot of otherwise needless work, and in all actuality, has necessitated having to open EACH AND EVERY folder that has been 'collected' in order to inspect the unchanged dates of the contents and then change the date(s) of the folder to properly reflect when the folder was actually created (and/or modified as that is what is reflected in the collating sequence displayed in the GUI.)
Can this be rectified or at least be made an option to retain the original date(s)?

**Comment by jbrodriguez** [comment]
That shouldn't be the case.

I haven't seen this kind of issue reported since I switched from diskmv to straight rsync, and the date/time subject was one of the reasons for the change.

Internally, unBALANCE uses rsync to copy files/folders and by default, rsync sets the original creation datetime when the copy has finished.

While it's copying it does set the creation datetime to 'now', but once the copy is done, it restores the original datetime.

Can you double check your observations ?

**Comment by Attidudinal** [comment]
What I've presented is absolutely correct - I have dozens upon dozens of folders whose contents were merged by the 'gather' process (in order to correct a "split" made by unRAID). ALL dates were changed to the date of operation; the modified date, the accessed data, AND the creation date - despite the fact that the folder already existed.
The folder contents reveal the original date the folder was created.  By looking at the computer-generated metadata contents this can be determined.
A representative sample of one of these folders and its contents has been attached.
This was modified by unBalance v2018.06.05, the most current version, that was run across the network and connected to with Web-UI v5.2.1 using https.

[Sample.docx](https://github.com/jbrodriguez/unbalance/files/2195032/Sample.docx)

**Comment by jbrodriguez** [comment]
unBALANCE uses rsync internally, invoked by default with "-avPR" flags.

The "a" flag translates to "rlptgoD", "t" being the relevant in this matter: preserve modification times.

So, while modification time is preserved, creation time is set to the date of the operation, unless the folder is present (I guess).

That's standard rsync behavior and there's no way around it that I can see.

**Comment by Attidudinal** [comment]
I won't dispute that there may be some shortcomings in the handling of time by a the function - a cursory overview of the times affected by the function would include atime, mtime, and ctime.
Apparently, the functioning of atime has been disputed and a number of more specific function calls were created to handle the case of 'inappropriate' assignments of time.
A description of ctime states that it "used to be" creation time, but does not definitively state just what it is now, other than to say that it cannot be set to an 'arbitrary' value.
The determinant would therefore be what is it that triggers such a change in the creation time when the folder exists (creation time is obviously the time of operation otherwise)...
This I cannot determine myself as I no longer code and when I did, it did not pertain to Unix/Linux, although my experience has taught me to test such factors in operation, despite what the documentation may say.  I have no doubt that this would have to be attempted to see just why it is that the creation date is being changed on an existing folder - the files themselves that are being merged within that folder aren't an issue as their timestamps (all of them) are intact after the operation.
If it appears there is no effective means of correcting the dysfunction of the rsync function (short of recoding it, or providing your own coding - as has been proven to be the case with atime), then
I can think of a very easy method of circumventing the dysfunction of rsync - since the folder that is being merged into (gather'ed) already exists, it should be relatively easy to capture the current creation timestamp of the folder and then after rsync has done its job of merging files, to update the creation timestamp of the folder with the saved value.  This could be done in a manner similar to the 'touch' program, except that the inputs of utime that it utilizes would be passed from the saved value instead of being input.
What do you think?

Reference sources regarding timestamps used by rsync and ctime usage...
https://download.samba.org/pub/rsync/rsync.html
https://en.wikipedia.org/wiki/Stat_(system_call)#ctime

Note: As stated previously in the presentation of the situation, NONE of the existing folder timestamps (access time, modification time, and creation time) are maintained; ALL have been altered.  And, while such modification may be debatable in the first two instances, the alteration of the created time for an existing folder is obviously in error. ('Creating' a number of repercussions.)
I also question the change of modification time (which isn't being maintained) only because it is the primary folder timestamp being is utilized by the Web-GUI (and Windows).
But then if the creation time is maintained, then there is at least a reference for the change of the modified timestamp if desired...
Do you happen to know of a program that will automate the update of a timestamp for a number of folders based on such a reference?

**Comment by jbrodriguez** [comment]
Closing this one, feel free to add any comments.

**Comment by Taipan24** [comment]
> Do you happen to know of a program that will automate the update of a timestamp for a number of folders based on such a reference?

Yes...Total Commander (a Windows app). If you have a copy of the original folders with the creation date/time (I used a backup for that), after using unbalance you can use the asymmetric "Synchronize Dirs" function in Total Commander to compare both folders, Ctrl-A to mark all folders/files, right click and select "copy file properties --> (left to right)" to copy properties from the backup folders to the folders created by unbalance.

I found this thread for the same reason: folder creation dates changed after using unbalance.
